clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/MigrateServerGroupStrategy.java

```diff
@@ -0,0 +1,271 @@
+   * @param migrateSecurityGroupStrategy the security group migration strategy
+   * @param subnetType                   the subnetType in which to migrate the server group (should be null for EC Classic migrations)
+   * @param iamRole                      the iamRole to use when migrating (optional)
+   * @param keyPair                      the keyPair to use when migrating (optional)
+   * @param targetAmi                    the target imageId to use when migrating (optional)
+   * @param dryRun                       whether to perform the migration or simply calculate the migration
+   * @return a result set indicating the components required to perform the migration (if a dry run), or the objects
+   * updated by the migration (if not a dry run)
+   */
+  public MigrateServerGroupResult generateResults(ServerGroupLocation source, ServerGroupLocation target,
+                                                  SecurityGroupLookup sourceLookup, SecurityGroupLookup targetLookup,
+                                                  MigrateLoadBalancerStrategy migrateLoadBalancerStrategy,
+                                                  MigrateSecurityGroupStrategy migrateSecurityGroupStrategy,
+                                                  String subnetType, String iamRole, String keyPair, String targetAmi,
+                                                  boolean dryRun) {
+
+    this.sourceLookup = sourceLookup;
+    this.targetLookup = targetLookup;
+    this.migrateSecurityGroupStrategy = migrateSecurityGroupStrategy;
+    this.getMigrateLoadBalancerStrategy = migrateLoadBalancerStrategy;
+
+    AsgService asgService = getRegionScopedProviderFactory().forRegion(source.getCredentials(), source.getRegion())
+      .getAsgService();
+
+    AutoScalingGroup sourceGroup = asgService.getAutoScalingGroup(source.getName());
+
+    if (sourceGroup == null) {
+      throw new IllegalStateException("Error retrieving source server group: " + source.getName());
+    }
+
+    LaunchConfiguration launchConfig = asgService.getLaunchConfiguration(sourceGroup.getLaunchConfigurationName());
+
+    if (launchConfig == null) {
+      throw new IllegalStateException("Could not find launch config: " + sourceGroup.getLaunchConfigurationName());
+    }
+
+    Names names = Names.parseName(source.getName());
+
+    List<MigrateLoadBalancerResult> targetLoadBalancers = generateTargetLoadBalancers(
+      sourceGroup, source, target, subnetType, dryRun);
+
+    MigrateServerGroupResult migrateResult = new MigrateServerGroupResult();
+
+    List<MigrateSecurityGroupResult> targetSecurityGroups = generateTargetSecurityGroups(launchConfig, source,  target,
+      migrateResult, dryRun);
+
+    Map<String, List<String>> zones = new HashMap<>();
+    zones.put(target.getRegion(), target.getAvailabilityZones());
+
+    DeploymentResult result;
+    if (!dryRun) {
+      Capacity capacity = getCapacity();
```

Comment:
There's a ton of duplication, but I think it's deceptively difficult to refactor.